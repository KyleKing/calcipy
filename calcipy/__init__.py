"""calcipy."""

from loguru import logger

__version__ = '0.21.5'
__pkg_name__ = 'calcipy'

logger.disable(__pkg_name__)

# ====== Above is the recommended code from calcipy_template and may be updated on new releases ======

# Load entry point for CLI
from .cli.main import run  # noqa: E402

__all__ = ('run',)
