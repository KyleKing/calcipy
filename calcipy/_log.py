"""Internal log instance."""

from shoal.log import get_logger

logger = get_logger()
