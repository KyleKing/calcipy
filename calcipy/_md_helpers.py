"""Markdown utilities."""

from beartype.typing import Any, Dict, List


def _format_md_table(headers: List[str], records: List[Dict[str, Any]]) -> List[str]:
    """Format the input as a Github markdown table."""
    table = [[str(_r[col]) for col in headers] for _r in records]
    widths = [max(len(row[col_idx].strip()) for row in [headers, *table]) for col_idx in range(len(headers))]

    def pad(values: List[str]) -> List[str]:
        return [val.strip().ljust(widths[col_idx]) for col_idx, val in enumerate(values)]

    def join(row: List[str], spacer: str = ' ') -> str:
        return f'|{spacer}' + f'{spacer}|{spacer}'.join(row) + f'{spacer}|'

    return [
        join(pad(headers)),
        join(['-' * widths[col_idx] for col_idx in range(len(headers))], '-'),
        *[join(pad(row)) for row in table],
    ]
