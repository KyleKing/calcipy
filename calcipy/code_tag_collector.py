"""Collect code tags and output for review in a single location."""

import re
from collections import defaultdict
from functools import lru_cache
from pathlib import Path
from subprocess import CalledProcessError  # nosec

import pandas as pd
import pendulum
from attrs import field, frozen
from attrs_strict import type_validator
from beartype import beartype
from beartype.typing import Dict, List, Optional, Pattern, Sequence, Tuple
from loguru import logger

from .file_helpers import read_lines
from .log_helpers import log_fun
from .proc_helpers import run_cmd

SKIP_PHRASE = 'calcipy:skip_tags'
"""String that indicates the file should be excluded from the tag search."""

COMMON_CODE_TAGS = ['FIXME', 'TODO', 'PLANNED', 'HACK', 'REVIEW', 'TBD', 'DEBUG']  # noqa: T100,T101,T103
"""Most common code tags. FYI and NOTE are excluded to not be tracked in the Code Summary."""

CODE_TAG_RE = r'((^|\s|\(|"|\')(?P<tag>{tag})(:| -)([^\r\n]))(?P<text>.+)'
"""Default code tag regex with `tag` and `text` matching groups.

Requires formatting with list of tags: `CODE_TAG_RE.format(tag='|'.join(tag_list))`

Commonly, the `tag_list` could be `COMMON_CODE_TAGS`

"""


@frozen
class _CodeTag:  # noqa: H601
    """Code Tag (FIXME,TODO,etc) with contextual information."""  # noqa: T100,T101

    lineno: int = field(validator=type_validator())
    tag: str = field(validator=type_validator())
    text: str = field(validator=type_validator())


@frozen
class _Tags:  # noqa: H601
    """Collection of code tags with additional contextual information."""

    path_source: Path = field(validator=type_validator())
    code_tags: List[_CodeTag] = field(validator=type_validator())


@beartype
def _search_lines(
    lines: List[str], regex_compiled: Pattern[str],
    skip_phrase: str = 'calcipy:skip_tags',
) -> List[_CodeTag]:
    """Search lines of text for matches to the compiled regular expression.

    Args:
        lines: lines of text as list
        regex_compiled: compiled regular expression. Expected to have matching groups `(tag, text)`
        skip_phrase: skip file if string is found in final two lines. Default is `SKIP_PHRASE`

    Returns:
        List[_CodeTag]: list of all code tags found in lines

    """
    if skip_phrase in '\n'.join(lines[-2:]):
        return []

    comments = []
    for lineno, line in enumerate(lines):
        match = regex_compiled.search(line)
        if match:
            group = match.groupdict()
            comments.append(_CodeTag(lineno + 1, tag=group['tag'], text=group['text']))
    return comments


@beartype
def _search_files(paths_source: Sequence[Path], regex_compiled: Pattern[str]) -> List[_Tags]:
    """Collect matches from multiple files.

    Args:
        paths_source: list of source files to parse
        regex_compiled: compiled regular expression. Expected to have matching groups `(tag, text)`

    Returns:
        List[_Tags]: list of all code tags found in files

    """
    matches = []
    for path_source in paths_source:
        lines = []
        try:
            lines = read_lines(path_source)
        except UnicodeDecodeError as err:
            logger.debug('Could not parse: {path_source}', path_source=path_source, err=err)

        comments = _search_lines(lines, regex_compiled)
        if comments:
            matches.append(_Tags(path_source, comments))

    return matches


@lru_cache(maxsize=128)
@beartype
def _git_info(cwd: Path) -> Tuple[Path, str]:
    """Collect information about the local git repository.

    Based on snippets from: https://gist.github.com/abackstrom/4034721#gistcomment-3982270
    and: https://github.com/rscherf/GitLink/blob/e2e7c412630246efc86de4fe71192f15bf11209e/GitLink.py

    Args:
        cwd: Path to the current working directory (typically file_path.parent)

    Returns:
        Tuple[Path, str]: (git_dir, repo_url)

    """
    git_dir = Path(run_cmd('git rev-parse --show-toplevel', cwd=str(cwd)))
    clone_uri = run_cmd('git remote get-url origin', cwd=str(cwd))
    # Could be ssh or http (with or without .git)
    # git@github.com:KyleKing/calcipy.git
    # https://github.com/KyleKing/calcipy.git
    sub_url = re.findall(r'^.+github.com[:/]([^.]+)(?:\.git)?$', clone_uri)[0]
    repo_url = f'https://github.com/{sub_url}'
    return (git_dir, repo_url)


@beartype
def _format_record(base_dir: Path, file_path: Path, comment: _CodeTag) -> Dict[str, str]:
    """Format each table row for the code tag summary file. Include git permalink.

    Args:
        base_dir: base path of the project if git directory is not known
        file_path: path to the file of interest
        comment: _CodeTag information for the matched tag

    Returns:
        Dict[str, str]: formatted dictionary with file info

    """
    cwd = file_path.parent
    git_dir, repo_url = _git_info(cwd=cwd)
    blame = None
    try:
        blame = run_cmd(f'git blame {file_path} -L {comment.lineno},{comment.lineno} --porcelain', cwd=cwd)
    except CalledProcessError:
        logger.exception('Failed to locate {file_path}', file_path=file_path)
    # Set fallback values if git logic doesn't work
    rel_path = file_path.relative_to(base_dir)
    source_file = f'{rel_path.as_posix()}:{comment.lineno}'
    ts = 'N/A'
    if blame:
        # Note: line number may be different in older blame (and relative path)
        revision, old_line_number = blame.split('\n')[0].split(' ')[:2]
        # If the change has not yet been committed, use the branch name as best guess
        if all(_c == '0' for _c in revision):
            revision = run_cmd('git branch --show-current', cwd=cwd)
        # Format a nice timestamp of the last edit to the line
        blame_dict = {
            line.split(' ')[0]: ' '.join(line.split(' ')[1:])
            for line in blame.split('\n')
        }
        # Handle uncommitted files that only have author-time and author-tz
        user = 'committer' if 'committer-tz' in blame_dict else 'author'
        dt = pendulum.from_timestamp(int(blame_dict[f'{user}-time']))
        tz = blame_dict[f'{user}-tz'][:3] + ':' + blame_dict[f'{user}-tz'][-2:]
        ts = pendulum.parse(dt.isoformat()[:-6] + tz).format('YYYY-MM-DD')
        # Filename may not be present if uncommitted. Use local path as fallback
        remote_file_path = blame_dict.get('filename', rel_path.as_posix())
        # PLANNED: Consider making "blame" configurable
        git_url = f'{repo_url}/blame/{revision}/{remote_file_path}#L{old_line_number}'
        source_file = f'[{source_file}]({git_url})'

    return {
        'Type': f'{comment.tag:>7}',
        'Comment': comment.text,
        'Last Edit': ts,
        'Source File': source_file,
    }


@beartype
def _format_report(
    base_dir: Path, code_tags: List[_Tags], tag_order: List[str],
) -> str:  # noqa: CCR001
    """Pretty-format the code tags by file and line number.

    Args:
        base_dir: base directory relative to the searched files
        code_tags: list of all code tags found in files
        tag_order: subset of all tags to include in the report and specified order

    Returns:
        str: pretty-formatted text

    """
    output = ''
    records = []
    counter: Dict[str, int] = defaultdict(lambda: 0)
    for comments in sorted(code_tags, key=lambda tc: tc.path_source, reverse=False):
        for comment in comments.code_tags:
            if comment.tag in tag_order:
                records.append(_format_record(base_dir, comments.path_source, comment))
                counter[comment.tag] += 1
    if records:
        df_tags = pd.DataFrame(records)
        output += df_tags.to_markdown(index=False)
    logger.debug('counter={counter}', counter=counter)

    sorted_counter = {tag: counter[tag] for tag in tag_order if tag in counter}
    logger.debug('sorted_counter={sorted_counter}', sorted_counter=sorted_counter)
    formatted_summary = ', '.join(f'{tag} ({count})' for tag, count in sorted_counter.items())
    if formatted_summary:
        output += f'\n\nFound code tags for {formatted_summary}\n'
    return output


@log_fun
@beartype
def write_code_tag_file(
    path_tag_summary: Path, paths_source: List[Path], base_dir: Path,
    regex_compiled: Optional[Pattern[str]] = None, tag_order: Optional[List[str]] = None,
    header: str = '# Task Summary',
) -> None:  # noqa: CCR001
    """Create the code tag summary file.

    Args:
        path_tag_summary: Path to the output file
        paths_source: list of source files to parse
        base_dir: base directory relative to the searched files
        regex_compiled: compiled regular expression. Expected to have matching groups `(tag, text)`.
            Default is CODE_TAG_RE with tags from tag_order
        tag_order: subset of all tags to include in the report and specified order. Default is COMMON_CODE_TAGS
        header: optional header text. Default is '# Task Summary'

    """
    tag_order = tag_order or COMMON_CODE_TAGS
    regex_compiled = regex_compiled or re.compile(CODE_TAG_RE.format(tag='|'.join(tag_order)))

    matches = _search_files(paths_source, regex_compiled)
    report = _format_report(base_dir, matches, tag_order=tag_order).strip()

    if report:
        path_tag_summary.write_text(f'{header}\n\n{report}\n\n<!-- {SKIP_PHRASE} -->\n')
    elif path_tag_summary.is_file():
        path_tag_summary.unlink()
