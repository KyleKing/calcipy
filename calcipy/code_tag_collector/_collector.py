"""Collect code tags and output for review in a single location."""

import re
from collections import defaultdict
from contextlib import suppress
from functools import lru_cache
from pathlib import Path
from subprocess import CalledProcessError  # nosec

import arrow
import pandas as pd
from beartype import beartype
from beartype.typing import Dict, List, Pattern, Sequence, Tuple
from corallium.file_helpers import read_lines
from corallium.log import logger
from corallium.shell import capture_shell
from pydantic import BaseModel

SKIP_PHRASE = 'calcipy_skip_tags'
"""String that indicates the file should be excluded from the tag search."""

COMMON_CODE_TAGS = ['FIXME', 'TODO', 'PLANNED', 'HACK', 'REVIEW', 'TBD', 'DEBUG']
"""Most common code tags.

FYI and NOTE are excluded to not be tracked in the Code Summary.

"""

CODE_TAG_RE = r'((^|\s|\(|"|\')(?P<tag>{tag})(:| -)([^\r\n]))(?P<text>.+)'
"""Default code tag regex with `tag` and `text` matching groups.

Requires formatting with list of tags: `CODE_TAG_RE.format(tag='|'.join(tag_list))`

Commonly, the `tag_list` could be `COMMON_CODE_TAGS`

"""


class _CodeTag(BaseModel):
    """Code Tag (FIXME,TODO,etc) with contextual information."""

    lineno: int
    tag: str
    text: str  # noqa: CCE001

    class Config:
        frozen = True


class _Tags(BaseModel):
    """Collection of code tags with additional contextual information."""

    path_source: Path
    code_tags: List[_CodeTag]  # noqa: CCE001

    class Config:
        frozen = True


@beartype
def _search_lines(
    lines: List[str], regex_compiled: Pattern[str],
    skip_phrase: str = 'calcipy_skip_tags',
) -> List[_CodeTag]:
    """Search lines of text for matches to the compiled regular expression.

    Args:
        lines: lines of text as list
        regex_compiled: compiled regular expression. Expected to have matching groups `(tag, text)`
        skip_phrase: skip file if string is found in final two lines. Default is `SKIP_PHRASE`

    Returns:
        List[_CodeTag]: list of all code tags found in lines

    """
    if skip_phrase in '\n'.join(lines[-2:]):
        return []

    max_len = 400
    comments = []
    for lineno, line in enumerate(lines):
        if match := regex_compiled.search(line):
            if len(line) <= max_len:  # FYI: Suppress long lines
                group = match.groupdict()
                comments.append(_CodeTag(lineno=lineno + 1, tag=group['tag'], text=group['text']))
            else:
                logger.text_debug('Skipping long line', lineno=lineno, line=line[:200])
    return comments


@beartype
def _search_files(paths_source: Sequence[Path], regex_compiled: Pattern[str]) -> List[_Tags]:
    """Collect matches from multiple files.

    Args:
        paths_source: list of source files to parse
        regex_compiled: compiled regular expression. Expected to have matching groups `(tag, text)`

    Returns:
        List[_Tags]: list of all code tags found in files

    """
    matches = []
    for path_source in paths_source:
        lines = []
        try:
            lines = read_lines(path_source)
        except UnicodeDecodeError as err:
            logger.text_debug('Could not parse', path_source=path_source, err=err)

        if comments := _search_lines(lines, regex_compiled):
            matches.append(_Tags(path_source=path_source, code_tags=comments))

    return matches


@lru_cache(maxsize=128)
@beartype
def _git_info(cwd: Path) -> Tuple[Path, str]:
    """Collect information about the local git repository.

    Based on snippets from: https://gist.github.com/abackstrom/4034721#gistcomment-3982270
    and: https://github.com/rscherf/GitLink/blob/e2e7c412630246efc86de4fe71192f15bf11209e/GitLink.py

    Args:
        cwd: Path to the current working directory (typically file_path.parent)

    Returns:
        Tuple[Path, str]: (git_dir, repo_url)

    """
    git_dir = Path(capture_shell('git rev-parse --show-toplevel', cwd=cwd))
    clone_uri = 'https://github.com/'
    with suppress(CalledProcessError):
        clone_uri = capture_shell('git remote get-url origin', cwd=cwd)
    # Could be ssh or http (with or without .git)
    # > git@github.com:KyleKing/calcipy.git
    # > https://github.com/KyleKing/calcipy.git
    sub_url = re.findall(r'^.+github.com[:/]([^.]+)(?:\.git)?$', clone_uri)[0]
    repo_url = f'https://github.com/{sub_url}'
    return git_dir, repo_url


class _CollectorRow(BaseModel):
    """Each row of the Code Tag table."""

    tag_name: str
    comment: str
    last_edit: str
    source_file: str

    @classmethod
    @beartype
    def from_code_tag(cls, code_tag: _CodeTag, last_edit: str, source_file: str) -> '_CollectorRow':
        return cls(
            tag_name=f'{code_tag.tag:>7}',
            comment=code_tag.text,
            last_edit=last_edit,
            source_file=source_file,
        )


@beartype
def _format_from_blame(
    *, collector_row: _CollectorRow, blame: str, repo_url: str, cwd: Path, rel_path: Path,
) -> _CollectorRow:
    """Parse the git blame for useful timestamps and author when available."""
    # Note: line number may be different in older blame (and relative path)
    revision, old_line_number = blame.split('\n')[0].split(' ')[:2]
    # If the change has not yet been committed, use the branch name as best guess
    if all(_c == '0' for _c in revision):
        revision = capture_shell('git branch --show-current', cwd=cwd)
    # Format a nice timestamp of the last edit to the line
    blame_dict = {
        line.split(' ')[0]: ' '.join(line.split(' ')[1:])
        for line in blame.split('\n')
    }

    # Handle uncommitted files that only have author-time and author-tz
    user = 'committer' if 'committer-tz' in blame_dict else 'author'
    dt = arrow.get(int(blame_dict[f'{user}-time']))
    tz = blame_dict[f'{user}-tz'][:3] + ':' + blame_dict[f'{user}-tz'][-2:]
    collector_row.last_edit = arrow.get(dt.isoformat()[:-6] + tz).format('YYYY-MM-DD')

    # Filename may not be present if uncommitted. Use local path as fallback
    remote_file_path = blame_dict.get('filename', rel_path.as_posix())
    # Assumes Github format
    git_url = f'{repo_url}/blame/{revision}/{remote_file_path}#L{old_line_number}'
    collector_row.source_file = f'[{collector_row.source_file}]({git_url})'

    return collector_row


@beartype
def _format_record(base_dir: Path, file_path: Path, comment: _CodeTag) -> _CollectorRow:
    """Format each table row for the code tag summary file. Include git permalink.

    Args:
        base_dir: base path of the project if git directory is not known
        file_path: path to the file of interest
        comment: _CodeTag information for the matched tag

    Returns:
        Dict[str, str]: formatted dictionary with file info

    """
    cwd = file_path.parent
    _git_dir, repo_url = _git_info(cwd=cwd)

    # Set fallback values if git logic doesn't work
    rel_path = file_path.relative_to(base_dir)
    collector_row = _CollectorRow.from_code_tag(
        code_tag=comment,
        last_edit='N/A',
        source_file=f'{rel_path.as_posix()}:{comment.lineno}',
    )

    try:
        blame = capture_shell(f'git blame {file_path} -L {comment.lineno},{comment.lineno} --porcelain', cwd=cwd)
        collector_row = _format_from_blame(
            collector_row=collector_row, blame=blame, repo_url=repo_url, cwd=cwd, rel_path=rel_path,
        )
    except CalledProcessError as exc:
        handled_errors = (128,)
        if exc.returncode not in handled_errors:
            raise
        logger.text_debug('Skipping blame', file_path=file_path, exc=exc)

    return collector_row


@beartype
def _format_report(  # noqa: CAC001
    base_dir: Path, code_tags: List[_Tags], tag_order: List[str],
) -> str:
    """Pretty-format the code tags by file and line number.

    Args:
        base_dir: base directory relative to the searched files
        code_tags: list of all code tags found in files
        tag_order: subset of all tags to include in the report and specified order

    Returns:
        str: pretty-formatted text

    """
    output = ''
    records = []
    counter: Dict[str, int] = defaultdict(lambda: 0)
    for comments in sorted(code_tags, key=lambda tc: tc.path_source, reverse=False):
        for comment in comments.code_tags:
            if comment.tag in tag_order:
                collector_row = _format_record(base_dir, comments.path_source, comment)
                records.append({
                    'Type': collector_row.tag_name,
                    'Comment': collector_row.comment,
                    'Last Edit': collector_row.last_edit,
                    'Source File': collector_row.source_file,
                })
                counter[comment.tag] += 1
    if records:
        df_tags = pd.DataFrame(records)
        # Prevent URLs from appearing on multiple lines
        content = df_tags.to_markdown(index=False, tablefmt='github', maxcolwidths=None) or ''
        for line in content.split('\n'):
            if not line.startswith('/'):
                output += '\n'
            output += line
    logger.text_debug('counter', counter=counter)

    sorted_counter = {tag: counter[tag] for tag in tag_order if tag in counter}
    logger.text_debug('sorted_counter', sorted_counter=sorted_counter)
    if formatted_summary := ', '.join(
        f'{tag} ({count})' for tag, count in sorted_counter.items()
    ):
        output += f'\n\nFound code tags for {formatted_summary}\n'
    return output


@beartype
def write_code_tag_file(
    path_tag_summary: Path,
    paths_source: List[Path],
    base_dir: Path,
    regex: str = '',
    tags: str = '',
    header: str = '# Task Summary\n\nAuto-Generated by `calcipy`',
) -> None:
    """Create the code tag summary file.

    Args:
        path_tag_summary: Path to the output file
        paths_source: list of source files to parse
        base_dir: base directory relative to the searched files
        regex: compiled regular expression. Expected to have matching groups `(tag, text)`.
            Default is CODE_TAG_RE with tags from tag_order
        tags: subset of all tags to include in the report and specified order. Default is COMMON_CODE_TAGS
        header: header text

    """
    tag_order = [_t.strip() for _t in tags.split(',') if _t] or COMMON_CODE_TAGS
    regex_compiled = re.compile((regex or CODE_TAG_RE).format(tag='|'.join(tag_order)))

    matches = _search_files(paths_source, regex_compiled)
    if report := _format_report(
        base_dir, matches, tag_order=tag_order,
    ).strip():
        path_tag_summary.write_text(f'{header}\n\n{report}\n\n<!-- {SKIP_PHRASE} -->\n')
        logger.text('Created Code Tag Summary', path_tag_summary=path_tag_summary)
    elif path_tag_summary.is_file():
        path_tag_summary.unlink()
