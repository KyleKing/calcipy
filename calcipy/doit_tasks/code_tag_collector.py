"""Collect code tags and output for review in a single location."""

from collections import defaultdict
from pathlib import Path
from typing import Dict, List, Pattern, Sequence

import attr
from beartype import beartype
from loguru import logger

from ..file_helpers import read_lines
from ..log_helpers import log_fun
from .base import debug_task
from .doit_globals import DG, DoitTask

SKIP_PHRASE = 'calcipy:skip_tags'
"""String that indicates the file should be excluded from the tag search."""


@attr.s(auto_attribs=True)
class _CodeTag:  # noqa: H601
    """Code Tag (FIXME,TODO,etc) with contextual information."""  # noqa: T100,T101

    lineno: int
    tag: str
    text: str


@attr.s(auto_attribs=True)
class _Tags:  # noqa: H601
    """Collection of code tags with additional contextual information."""

    path_source: Path
    code_tags: List[_CodeTag]


@beartype
def _search_lines(
    lines: List[str], regex_compiled: Pattern[str],
    skip_phrase: str = 'calcipy:skip_tags',
) -> List[_CodeTag]:
    """Search lines of text for matches to the compiled regular expression.

    Args:
        lines: lines of text as list
        regex_compiled: compiled regular expression. Expected to have matching groups `(tag, text)`
        skip_phrase: skip file if string is found in final two lines. Default is `SKIP_PHRASE`

    Returns:
        List[_CodeTag]: list of all code tags found in lines

    """
    if skip_phrase in '\n'.join(lines[-2:]):
        return []

    comments = []
    for lineno, line in enumerate(lines):
        match = regex_compiled.search(line)
        if match:
            mg = match.groupdict()
            comments.append(_CodeTag(lineno + 1, tag=mg['tag'], text=mg['text']))
    return comments


@beartype
def _search_files(paths_source: Sequence[Path], regex_compiled: Pattern[str]) -> List[_Tags]:
    """Collect matches from multiple files.

    Args:
        paths_source: list of source files to parse
        regex_compiled: compiled regular expression. Expected to have matching groups `(tag, text)`

    Returns:
        List[_Tags]: list of all code tags found in files

    """
    matches = []
    for path_source in paths_source:
        lines = []
        try:
            lines = read_lines(path_source)
        except UnicodeDecodeError as err:
            logger.warning(f'Could not parse: {path_source}', err=err)

        comments = _search_lines(lines, regex_compiled)
        if comments:
            matches.append(_Tags(path_source, comments))

    return matches


@beartype
def _format_report(base_dir: Path, code_tags: List[_Tags]) -> str:  # noqa: CCR001
    """Pretty-format the code tags by file and line number.

    Args:
        base_dir: base directory relative to the searched files
        code_tags: list of all code tags found in files

    Returns:
        str: pretty-formatted text

    """
    output = ''
    counter: Dict[str, int] = defaultdict(lambda: 0)
    for comments in sorted(code_tags, key=lambda tc: tc.path_source, reverse=False):
        output += f'- {comments.path_source.relative_to(base_dir).as_posix()}\n'
        for comment in comments.code_tags:
            output += f'    - line {comment.lineno:>3} {comment.tag:>7}: {comment.text}\n'
            counter[comment.tag] += 1
        output += '\n'
    logger.debug('counter={counter}', counter=counter)

    sorted_counter = {tag: counter[tag] for tag in DG.ct.tags if tag in counter}
    logger.debug('sorted_counter={sorted_counter}', sorted_counter=sorted_counter)
    formatted_summary = ', '.join(f'{tag} ({count})' for tag, count in sorted_counter.items())
    if formatted_summary:
        output += f'Found code tags for {formatted_summary}\n'
    return output


@log_fun
@beartype
def _write_code_tag_file(path_tag_summary: Path) -> None:
    """Create the code tag summary file.

    Args:
        path_tag_summary: Path to the output file

    """
    header = f'# Task Summary\n\nAuto-Generated by `{DG.meta.pkg_name}`'
    regex_compiled = DG.ct.compile_issue_regex()
    matches = _search_files(DG.meta.paths, regex_compiled)
    report = _format_report(DG.meta.path_project, matches).strip()
    if report:
        path_tag_summary.write_text(f'{header}\n\n{report}\n\n<!-- {SKIP_PHRASE} -->\n')
    elif path_tag_summary.is_file():
        path_tag_summary.unlink()


@beartype
def task_collect_code_tags() -> DoitTask:
    """Create a summary file with all of the found code tags.

    Returns:
        DoitTask: doit task

    """
    return debug_task([(_write_code_tag_file, (DG.ct.path_code_tag_summary,))])
