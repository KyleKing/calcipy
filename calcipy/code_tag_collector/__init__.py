try:
    from ._collector import write_code_tag_file
except ImportError as exc:
    raise RuntimeError("The 'calcipy[tags]' extras are missing") from exc
