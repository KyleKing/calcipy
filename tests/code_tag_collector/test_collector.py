import re

from calcipy.code_tag_collector import write_code_tag_file
from calcipy.code_tag_collector._collector import CODE_TAG_RE, _CodeTag, _format_report, _search_lines, _Tags

from ..configuration import TEST_DATA_DIR

TEST_PROJECT = TEST_DATA_DIR / 'test_project'


def test__search_lines():
    lines = [
        '# DEBUG: Show dodo.py in the documentation',  # noqa: T001
        'print("FIXME: Show README.md in the documentation (may need to update paths?)")',  # noqa: T100
        '# FYI: Replace src_examples_dir and make more generic to specify code to include in documentation',
        '# HACK: Show table of contents in __init__.py file',  # noqa: T103
        '# NOTE: Show table of contents in __init__.py file',
        '# PLANNED: Show table of contents in __init__.py file',
        '# REVIEW: Show table of contents in __init__.py file',
        '# TBD: Show table of contents in __init__.py file',
        '# TODO: Show table of contents in __init__.py file',  # noqa: T101
        '# HACK - Support unconventional dashed code tags',  # noqa: T101
        'class Code: # TODO: Complete',  # noqa: T101
        '   //TODO: Not matched',  # noqa: T101
        '   ...  # Both FIXME: and FYI: in the same line, but only match the first',  # noqa: T100,T101
        '# FIXME: ' + 'For a long line is ignored ...' * 14,  # noqa: T100,T101
    ]
    tag_order = ['FIXME', 'FYI', 'HACK', 'REVIEW']  # noqa: T100
    regex_compiled = re.compile(CODE_TAG_RE.format(tag='|'.join(tag_order)))

    comments = _search_lines(lines, regex_compiled)

    # TODO: assert_against_cache(comments)
    assert [_c.dict() for _c in comments] == [
        {
            'lineno': 2,
            'tag': 'FIXME',
            'text': 'Show README.md in the documentation (may need to update paths?)\")',
        },
        {
            'lineno': 3,
            'tag': 'FYI',
            'text': 'Replace src_examples_dir and make more generic to specify code to include in documentation',
        },
        {
            'lineno': 4,
            'tag': 'HACK',
            'text': 'Show table of contents in __init__.py file',
        },
        {
            'lineno': 7,
            'tag': 'REVIEW',
            'text': 'Show table of contents in __init__.py file',
        },
        {
            'lineno': 10,
            'tag': 'HACK',
            'text': 'Support unconventional dashed code tags',
        },
        {
            'lineno': 13,
            'tag': 'FIXME',
            'text': 'and FYI: in the same line, but only match the first',
        },
    ]


def test__format_report(fake_process):
    fake_process.pass_command([fake_process.any()])  # Allow "git blame" and other commands to run unregistered
    fake_process.keep_last_process(keep=True)
    lines = ['# DEBUG: Example 1', '# TODO: Example 2']  # noqa: T101
    comments = [
        _CodeTag(lineno=lineno, **dict(zip(('tag', 'text'), line.split('# ')[1].split(': '))))  # noqa: B905
        for lineno, line in enumerate(lines)
    ]
    tagged_collection = [_Tags(path_source=TEST_DATA_DIR / 'test_project', code_tags=comments)]
    tag_order = ['TODO']  # noqa: T101

    output = _format_report(TEST_DATA_DIR.parent, tagged_collection, tag_order=tag_order)

    # TODO: assert_against_cache({'output': output.split('\n')})
    assert output.split('\n') == [
        '',
        '| Type   | Comment   | Last Edit   | Source File         |',
        '|--------|-----------|-------------|---------------------|',
        '| TODO   | Example 2 | N/A         | data/test_project:1 |',
        '',
        'Found code tags for TODO (1)',
        '',
    ]


def test_write_code_tag_file_when_no_matches(fix_test_cache):
    path_tag_summary = fix_test_cache / 'code_tags.md'
    tmp_code_file = fix_test_cache / 'tmp.code'
    tmp_code_file.write_text('No FIXMES or TODOS here')

    write_code_tag_file(
        path_tag_summary=path_tag_summary, paths_source=[tmp_code_file], base_dir=fix_test_cache,
    )

    assert not path_tag_summary.is_file()


# calcipy_skip_tags
