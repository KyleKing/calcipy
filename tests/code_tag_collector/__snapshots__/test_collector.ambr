# serializer version: 1
# name: test__format_report
  '''
  
  | Type    | Comment   | Last Edit | Source File         |
  |---------|-----------|-----------|---------------------|
  |    TODO | Example 2 | N/A       | data/test_project:1 |
  
  Found code tags for TODO (1)
  
  '''
# ---
# name: test__search_lines
  list([
    _CodeTag(lineno=2, tag='FIXME', text='Show README.md in the documentation (may need to update paths?)")'),
    _CodeTag(lineno=3, tag='FYI', text='Replace src_examples_dir and make more generic to specify code to include in documentation'),
    _CodeTag(lineno=4, tag='HACK', text='Show table of contents in __init__.py file'),
    _CodeTag(lineno=7, tag='REVIEW', text='Show table of contents in __init__.py file'),
    _CodeTag(lineno=10, tag='HACK', text='Support unconventional dashed code tags'),
    _CodeTag(lineno=13, tag='FIXME', text='and FYI: in the same line, but only match the first'),
  ])
# ---
